test/test_testing.py

```diff
@@ -234,6 +234,34 @@ def test_metadata_from_directory(self):
                 db:
                     interface: pgsql
             '''))
+        harness = self._get_dummy_charm_harness(tmp)
+        harness.begin()
+        self.assertEqual(['db'], list(harness.model.relations))
```

Comment:
I think we've been using the expected value is the second one. This is a copy & paste from existing code, though, so it's fine to merge it as-is and address it in a follow up.

Reply:
We should be consistent on ordering. However, I do believe 'unittest.assertEqual' assumes (expected, actual) ordering. Because if you do:
```
self.assertEqual(['foo', 'bar', 'baz'], ['foo', 'bar', 'baz', 'bing'])
```
you get:
```
Traceback (most recent call last):
  File "/home/jameinel/dev/go/src/github.com/canonical/operator/test/test_testing.py", line 376, in test_order
    self.assertEqual(['foo', 'bar', 'baz'], ['foo', 'bar', 'baz', 'bing'])
AssertionError: Lists differ: ['foo', 'bar', 'baz'] != ['foo', 'bar', 'baz', 'bing']

Second list contains 1 additional elements.
First extra element 3:
'bing'

- ['foo', 'bar', 'baz']
+ ['foo', 'bar', 'baz', 'bing']
?                     ++++++++
```
And I realize that with any diff it can go in either order. But a '++' to *me* means that the actual content added something, not that it is missing because the expected content has it.

Also, PyCharm actually does:
```
['foo', 'bar', 'baz', 'bing'] != ['foo', 'bar', 'baz']

Expected :['foo', 'bar', 'baz']
Actual   :['foo', 'bar', 'baz', 'bing']
```

eg, they annotate the first param as "expected' and the second as "actual". I could accept that it is a PyCharm oddity, but it seems consistent with just plain './run_tests' as well.


I can learn to read the diffs backward and invert the meaning in my editor if we feel the other convention is more obvious.

Reply:
We should be consistent on ordering. However, I do believe 'unittest.assertEqual' assumes (expected, actual) ordering. Because if you do:
```
self.assertEqual(['foo', 'bar', 'baz'], ['foo', 'bar', 'baz', 'bing'])
```
you get:
```
Traceback (most recent call last):
  File "/home/jameinel/dev/go/src/github.com/canonical/operator/test/test_testing.py", line 376, in test_order
    self.assertEqual(['foo', 'bar', 'baz'], ['foo', 'bar', 'baz', 'bing'])
AssertionError: Lists differ: ['foo', 'bar', 'baz'] != ['foo', 'bar', 'baz', 'bing']

Second list contains 1 additional elements.
First extra element 3:
'bing'

- ['foo', 'bar', 'baz']
+ ['foo', 'bar', 'baz', 'bing']
?                     ++++++++
```
And I realize that with any diff it can go in either order. But a '++' to *me* means that the actual content added something, not that it is missing because the expected content has it.

Also, PyCharm actually does:
```
['foo', 'bar', 'baz', 'bing'] != ['foo', 'bar', 'baz']

Expected :['foo', 'bar', 'baz']
Actual   :['foo', 'bar', 'baz', 'bing']
```

eg, they annotate the first param as "expected' and the second as "actual". I could accept that it is a PyCharm oddity, but it seems consistent with just plain './run_tests' as well.


I can learn to read the diffs backward and invert the meaning in my editor if we feel the other convention is more obvious.

Reply:
Let's please not invert it.. as we talked last week, it's pretty much universal to have the constant as the second value. Almost no one writes code like
```
if "something" == myvar ...
```
If you look at the examples in the docstrings of Python's unittest itself, they are also clear:
```
            self.assertEqual(cm.output, ['INFO:foo:first message',
                                         'ERROR:foo.bar:second message'])
...
               self.assertEqual(the_exception.error_code, 3)
```
and so on.

I don't know what PyCharm does, but that's up to them.